### ISSUE TITLE
[RFC] add new dependency CI test

### ISSUE THREAD

**Issue by legalsylvain** [OCA/OpenUpgrade#3399]
Rational :  
- we should not merge migration script if dependant modules are not marked as migrated. 
Regarding that PR  https://github.com/OCA/OpenUpgrade/pull/3394
(website_forum migration) the CI is green. However, website is not marked as migrated on 15.0 head branch. (See https://github.com/OCA/OpenUpgrade/blob/15.0/docsource/modules140-150.rst)

Proposal : 
Add a new test name dependency-test : 
Get all the module to update. 
Get all the dependencies.
If one dependency is not in all the module to update, raise an error. 

This could make the life of the maintainers easier. 
For the time being, for trivial PR we have to :  
- list the dependencies
- check for each module, if the module is covered

What do you think : @OCA/openupgrade-maintainers ?

I think the problem get bigger in new versions because 90% of the PR are trivial / nothing to do.

As a résult trivial PR are not merged.

Note : alternative proposal could be to link the pending PR during the call of "ocabot migration" so we could have all the related PR linked together.

**Comment by pedrobaeza** [comment]
Yeah, excellent idea. I don't know right now if it's easy to be implemented, but the goal is important.

**Comment by MiquelRForgeFlow** [comment]
I ❤️ this idea.

**Comment by gdgellatly** [comment]
Hi,

Could we parse the file differently for modules to test? I mean committing a Nothing to Do PR is effectively a noop. We should IMO let them be merged even if dependencies aren't migrated. Just to lower the workload and get those ones through almost without touching.

As a result trivial PR are merged, but module upgrade not run until all dependencies are merged

Something like a list (via bash, actions, some other way) of modules
Basically something like
```
to_remove = set()
for mod in set(parsed_modules_to_upgrade):
  if not all(m for m in parse_manifest_deps(mod)):
    if not migration_scripts:
      to_remove.add(mod)
modules_to_test = parsed_modules_to_upgrade - to_remove
```
To me it just seems easier to get the trivial NTD ones gone early. Otherwise we get a test fail that needs dealing with.

**Comment by legalsylvain** [comment]
I'm not sure to understand. 
With your proposal, @gdgellatly, if a PR is about ``website_forum`` migration and if ``website`` is not migrated, ``modules_to_test`` will not contains ``website_forum`` (neither ``website`` of course).
So basically you are proposing to disable the CI ? Is it what we want ? because, even if some "nothing to do" seems trivial, corner case can occures and maybe make an update of ``website_forum`` will fail once tested.

Don't you think ?

**Comment by gdgellatly** [comment]
@legalsylvain you are correct, rarely someone will make an error, or rarely there may be some corner case. And it will be caught when CI fails and can simply be reverted. Reverting 1 rare error is less work IMO.

I am more proposing/suggesting disable CI on trivial nothing to do PR's *if dependencies aren't migrated* as a way to reduce human overhead. If you feel it creates more overhead, then ignore it. 

In fact I'd go as far to say that TRUSTED maintainers could create and merge NTD PR's immediately and use reviewer time better elsewhere. Indeed, I would imagine that many analysis files are so obvious that that many NTD updates could even be automated at start of process.

**Comment by pedrobaeza** [comment]
Graeme, that's not correct. A nothing to do on a dependent module can't never be merged without the dependencies. The CI will not be active for that module, as the dependencies are not marked as installed and no possible problem can be detected.

**Comment by gdgellatly** [comment]
Why? Surely it would run once the dependencies are migrated?

**Comment by gdgellatly** [comment]
Look, you guys know loads more than me. If it can't work because of some OpenUpgrade limitation then fair enough. If it a CI limitation, then well maybe there is a way to rethink. I'm just giving alternate thoughts on how we can reduce some of the wasteful workload we have, especially as @legalsylvain says, up to 90% of PR's are going to be like this in future.

**Comment by pedrobaeza** [comment]
It's not a CI limitation. It's a modular with hierarchy system limitation. We may change things in dependency modules that require to modify things in the dependent ones.

**Comment by gdgellatly** [comment]
Maybe an alternative then, at start of process could we automate all the Nothing to do PR's (by some automated file analysis) in some way and not merge them, but then run CI again, if green, automerge them when all dependencies are merged

**Comment by pedrobaeza** [comment]
Another no for me . Not having things in the analysis file doesn't mean there's nothing to do. Unfortunately, most of the times contributors (including me sometimes) just base the pull requests in this file. Automating this will make that this attention lowers even more and more issues post migrations exist.

**Comment by gdgellatly** [comment]
@pedrobaeza Wouldn't it be easier to check a proposed PR? I must admit, in reviewing, the thing I find hard now is having to have both the module and the PR open. Also can you give me an example of an empty file that required a migration script, so I can see how having no analysis file still means there is upgrade work?

**Comment by legalsylvain** [comment]
Thank you for your enthusiasm !

Well, in a very first sight, I think that it will be easier in ocabot project, but the problem is that it is specific to Openupgrade. Something like : 
- When maintainers call "ocabot migration xxx", 
-> check dependencies modules; xxx depends on yyy and zzz (***)
-> check modules migrations in the main "Migration from 14 to 15".
-> if some dependencies are not done ("checked") post a message with the according PR, (if exist). Something like : 

```
"You propose the migration for the module xxx but : 
- the yyy module is not ported (see unmerged PR #...)
- the zzz module is not ported. (no PR found). Could you propose a PR for that module ?"
```

-> otherwise, optional (but explicit)
```
"You propose the migration fo the module xxx that depends on yyy and zzz
- The yyy module has been ported here PR#
- the zzz module has been ported here PR#
So, this PR could be merged, once reviewed and CI is green"
```

We that design, if I recheck a pending PR, 6 monthes later, I can very easily know that it is mergeable or not. 

We could plan to work on it during the OCA code sprint days ? It could be great to have this feature implemented for the V16 release !


> @pedrobaeza Wouldn't it be easier to check a proposed PR? I must admit, in reviewing, the thing I find hard now is having to have both the module and the PR open. Also can you give me an example of an empty file that required a migration script, so I can see how having no analysis file still means there is upgrade work?

If I remember correctly, on V7, there was a change in the partner framework. (res.partner.adress stuff, it's old...) AFAIK, you have to call a specific function for each ``partner_id`` field present on each model.
(https://github.com/OCA/openupgradelib/blob/master/openupgradelib/openupgrade_70.py#L29)

Not sure, thought.


(***) is there an OCA tools to get the dependencies of the module, based on the name ? I think yes, but I don't remember...

**Comment by MiquelRForgeFlow** [comment]
> Also can you give me an example of an empty file that required a migration script

It doesn't come to my mind now an exactly case like this, but there have been several cases where some scripts were required to adapt some things that the analysis doesn't detect.

**Comment by pedrobaeza** [comment]
> Also can you give me an example of an empty file that required a migration script, so I can see how having no analysis file still means there is upgrade work?

There are several examples of changes that are not reflected on migration analysis files:

- Quants merge in v11.
- Change of meaning of selection field values.
- Transformation of website pages from BS3 to BS4 (and now we will need the same for v16 due to https://github.com/odoo/odoo/pull/95450)
- Integer value range that changes: https://github.com/OCA/OpenUpgrade/pull/2827
- ...

This can happen as well on a module with no analysis file. Besides, most of them can have analysis file, but you should decode the nothing to do on the changes, which is not easy to automate.

**Comment by pedrobaeza** [comment]
Closing as stalled.
